# DATASET

## Concatenate all Parquet files by years

In [1]:
import glob
import pandas as pd

In [2]:
def create_df_year(year):
    dfs = []
    # Read parquet files from directory /data/{year} 
    for fin in glob.glob(f'./data/{year}/*.parquet'):
      dfs.append(pd.read_parquet(fin))
    df = pd.concat(dfs)
    return df

def save_as_parquet(df,name = 'df'):
        return df.to_parquet(f'{name}.parquet.gzip',
              compression='gzip')

In [3]:
df_2022 = create_df_year(2022)

In [4]:
df_2022

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               1  2022-01-01 00:35:40   2022-01-01 00:53:29              2.0   
1               1  2022-01-01 00:33:43   2022-01-01 00:42:07              1.0   
2               2  2022-01-01 00:53:21   2022-01-01 01:02:19              1.0   
3               2  2022-01-01 00:25:21   2022-01-01 00:35:23              1.0   
4               2  2022-01-01 00:36:48   2022-01-01 01:14:20              1.0   
...           ...                  ...                   ...              ...   
3399544         2  2022-12-31 23:46:00   2023-01-01 00:11:00              NaN   
3399545         2  2022-12-31 23:13:24   2022-12-31 23:29:08              NaN   
3399546         2  2022-12-31 23:00:49   2022-12-31 23:26:57              NaN   
3399547         1  2022-12-31 23:02:50   2022-12-31 23:16:05              NaN   
3399548         2  2022-12-31 23:00:15   2022-12-31 23:22:56              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 3.80         1.0                  N           142   
1                 2.10         1.0                  N           236   
2                 0.97         1.0                  N           166   
3                 1.09         1.0                  N           114   
4                 4.30         1.0                  N            68   
...                ...         ...                ...           ...   
3399544          11.56         NaN               None            16   
3399545           5.06         NaN               None            75   
3399546          13.35         NaN               None           168   
3399547           0.00         NaN               None           238   
3399548           4.24         NaN               None            85   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                 236             1        14.50    3.0      0.5        3.65   
1                  42             1         8.00    0.5      0.5        4.00   
2                 166             1         7.50    0.5      0.5        1.76   
3                  68             2         8.00    0.5      0.5        0.00   
4                 163             1        23.50    0.5      0.5        3.00   
...               ...           ...          ...    ...      ...         ...   
3399544            36             0        39.55    0.0      0.5        8.21   
3399545            50             0        26.23    0.0      0.5        0.00   
3399546           197             0        47.73    0.0      0.5        9.85   
3399547           116             0        12.74    0.0      0.5        0.00   
3399548            37             0        19.60    0.0      0.5        4.00   

         tolls_amount  improvement_surcharge  total_amount  \
0                 0.0                    0.3         21.95   
1                 0.0                    0.3         13.30   
2                 0.0                    0.3         10.56   
3                 0.0                    0.3         11.80   
4                 0.0                    0.3         30.30   
...               ...                    ...           ...   
3399544           0.0                    1.0         49.26   
3399545           0.0                    1.0         30.23   
3399546           0.0                    1.0         59.08   
3399547           0.0                    1.0         16.74   
3399548           0.0                    1.0         25.10   

         congestion_surcharge  airport_fee  
0                         2.5          0.0  
1                         0.0          0.0  
2                         0.0          0.0  
3                         2.5          0.0  
4                         2.5          0.0  
...                       ...          ...  
3399544                   NaN          NaN  
3399545                   NaN          NaN  
3399546                   NaN          NaN  
3399547                   N

In [5]:
df_2021 = create_df_year(2021)

In [6]:
df_2021

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               1  2021-01-01 00:30:10   2021-01-01 00:36:12              1.0   
1               1  2021-01-01 00:51:20   2021-01-01 00:52:19              1.0   
2               1  2021-01-01 00:43:30   2021-01-01 01:11:06              1.0   
3               1  2021-01-01 00:15:48   2021-01-01 00:31:01              0.0   
4               2  2021-01-01 00:31:49   2021-01-01 00:48:21              1.0   
...           ...                  ...                   ...              ...   
3214364         2  2021-12-31 23:18:02   2021-12-31 23:32:19              NaN   
3214365         2  2021-12-31 23:46:00   2022-01-01 00:07:00              NaN   
3214366         2  2021-12-31 23:46:00   2022-01-01 00:04:00              NaN   
3214367         2  2021-12-31 23:53:00   2022-01-01 00:02:00              NaN   
3214368         2  2021-12-31 23:19:28   2021-12-31 23:26:30              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 2.10         1.0                  N           142   
1                 0.20         1.0                  N           238   
2                14.70         1.0                  N           132   
3                10.60         1.0                  N           138   
4                 4.94         1.0                  N            68   
...                ...         ...                ...           ...   
3214364           3.99         NaN               None           145   
3214365           5.96         NaN               None           236   
3214366           4.99         NaN               None            48   
3214367           1.36         NaN               None            75   
3214368           1.99         NaN               None           170   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                  43             2         8.00    3.0      0.5        0.00   
1                 151             2         3.00    0.5      0.5        0.00   
2                 165             1        42.00    0.5      0.5        8.65   
3                 132             1        29.00    0.5      0.5        6.05   
4                  33             1        16.50    0.5      0.5        4.06   
...               ...           ...          ...    ...      ...         ...   
3214364           107             0        17.42    0.0      0.5        5.80   
3214365           158             0        23.05    0.0      0.5        2.85   
3214366            13             0        20.97    0.0      0.5        5.23   
3214367            75             0        13.20    0.0      0.5        2.00   
3214368           237             0        13.20    0.0      0.5        2.63   

         tolls_amount  improvement_surcharge  total_amount  \
0                 0.0                    0.3         11.80   
1                 0.0                    0.3          4.30   
2                 0.0                    0.3         51.95   
3                 0.0                    0.3         36.35   
4                 0.0                    0.3         24.36   
...               ...                    ...           ...   
3214364           0.0                    0.3         26.52   
3214365           0.0                    0.3         29.20   
3214366           0.0                    0.3         29.50   
3214367           0.0                    0.3         16.00   
3214368           0.0                    0.3         19.13   

         congestion_surcharge  airport_fee  
0                         2.5          NaN  
1                         0.0          NaN  
2                         0.0          NaN  
3                         0.0          NaN  
4                         2.5          NaN  
...                       ...          ...  
3214364                   NaN          NaN  
3214365                   NaN          NaN  
3214366                   NaN          NaN  
3214367                   N

In [7]:
df_2020 = create_df_year(2020)

In [8]:
df_2020

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               1  2020-01-01 00:28:15   2020-01-01 00:33:03              1.0   
1               1  2020-01-01 00:35:39   2020-01-01 00:43:04              1.0   
2               1  2020-01-01 00:47:41   2020-01-01 00:53:52              1.0   
3               1  2020-01-01 00:55:23   2020-01-01 01:00:14              1.0   
4               2  2020-01-01 00:01:58   2020-01-01 00:04:16              1.0   
...           ...                  ...                   ...              ...   
1461893         2  2020-12-31 23:44:35   2021-01-01 00:01:22              NaN   
1461894         2  2020-12-31 23:41:36   2020-12-31 23:50:32              NaN   
1461895         2  2020-12-31 23:01:17   2020-12-31 23:40:37              NaN   
1461896         2  2020-12-31 23:31:29   2020-12-31 23:44:22              NaN   
1461897         2  2020-12-31 23:12:48   2020-12-31 23:24:51              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 1.20         1.0                  N           238   
1                 1.20         1.0                  N           239   
2                 0.60         1.0                  N           238   
3                 0.80         1.0                  N           238   
4                 0.00         1.0                  N           193   
...                ...         ...                ...           ...   
1461893           9.22         NaN               None           143   
1461894           4.79         NaN               None             4   
1461895          28.00         NaN               None           210   
1461896           7.08         NaN               None           148   
1461897           2.35         NaN               None            18   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                 239             1         6.00    3.0      0.5        1.47   
1                 238             1         7.00    3.0      0.5        1.50   
2                 238             1         6.00    3.0      0.5        1.00   
3                 151             1         5.50    0.5      0.5        1.36   
4                 193             2         3.50    0.5      0.5        0.00   
...               ...           ...          ...    ...      ...         ...   
1461893            20             0        32.49    0.0      0.5        8.16   
1461894           262             0        13.22    0.0      0.5        3.65   
1461895            78             0        69.31    0.0      0.5        2.75   
1461896           129             0        35.95    0.0      0.5        2.75   
1461897           235             0        17.09    0.0      0.5        2.75   

         tolls_amount  improvement_surcharge  total_amount  \
0                0.00                    0.3         11.27   
1                0.00                    0.3         12.30   
2                0.00                    0.3         10.80   
3                0.00                    0.3          8.16   
4                0.00                    0.3          4.80   
...               ...                    ...           ...   
1461893          0.00                    0.3         43.95   
1461894          0.00                    0.3         20.17   
1461895          6.12                    0.3         78.98   
1461896          0.00                    0.3         39.50   
1461897          0.00                    0.3         20.64   

         congestion_surcharge airport_fee  
0                         2.5        None  
1                         2.5        None  
2                         2.5        None  
3                         0.0        None  
4                         0.0        None  
...                       ...         ...  
1461893                   NaN         NaN  
1461894                   NaN         NaN  
1461895                   NaN         NaN  
1461896                   NaN        

In [10]:
df_2020.info()

<class 'pandas.core.frame.DataFrame'>
Index: 24649092 entries, 0 to 1461897
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee            obj

<hr>

**SUMMARY NO. OF ROWS:** <br>
1. `df_2022` : 39,656,098 rows <br>
2. `df_2021` : 30,904,308 rows <br>
3. `df_2020` : 24,649,092 rows

##  Get the first N rows randomly from each month
`*.parquet.gzip` files are still too large to upload into Github and take too much storage/bandwidth even with Git LFS. (about 1.5GB)

Hence, reducing the file size is the best solution by filtering the rows for this side project use. Here, we will use first `20,000` rows randomly from each month

### 2022 Dataset CSV

In [24]:
df = pd.read_parquet('2022.parquet.gzip').dropna(subset=['passenger_count', 'RatecodeID'], how='all')
df = df.groupby([df['tpep_pickup_datetime'].dt.month]).apply(lambda x: x.sample(20000)).reset_index(drop=True).sort_values(by='tpep_pickup_datetime')
df

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
190481         2  2002-10-22 14:25:38   2002-10-22 14:37:18              1.0   
183318         2  2002-10-25 00:16:07   2002-10-25 00:16:13              1.0   
220733         2  2008-12-31 23:19:56   2008-12-31 23:44:06              1.0   
5181           2  2022-01-01 00:03:22   2022-01-01 00:29:53              1.0   
9816           2  2022-01-01 00:06:34   2022-01-01 00:12:34              1.0   
...          ...                  ...                   ...              ...   
226656         2  2022-12-31 23:41:16   2022-12-31 23:46:05              1.0   
235960         2  2022-12-31 23:42:22   2023-01-01 00:02:04              1.0   
228111         2  2022-12-31 23:42:33   2022-12-31 23:46:36              1.0   
232115         2  2022-12-31 23:45:16   2023-01-01 00:18:35              1.0   
225948         2  2022-12-31 23:57:50   2023-01-01 00:05:39              1.0   

        trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
190481           1.39         1.0                  N           162   
183318           0.00         5.0                  N           243   
220733           2.76         1.0                  N           161   
5181             2.70         1.0                  N            43   
9816             2.47         1.0                  N           140   
...               ...         ...                ...           ...   
226656           0.71         1.0                  N           246   
235960           1.35         1.0                  N           161   
228111           0.57         1.0                  N            90   
232115           4.08         1.0                  N            68   
225948           1.77         1.0                  N           143   

        DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
190481           170             2          9.0    0.0      0.5        0.00   
183318           243             2         29.8    0.0      0.0        0.00   
220733           142             2         12.5    0.5      0.5        0.00   
5181             186             2         17.0    0.5      0.5        0.00   
9816             170             1          9.0    0.5      0.5        2.56   
...              ...           ...          ...    ...      ...         ...   
226656            48             2          6.5    1.0      0.5        0.00   
235960           186             1         15.6    1.0      0.5        4.12   
228111           158             1          5.8    1.0      0.5        2.16   
232115           140             2         30.3    1.0      0.5        0.00   
225948           236             1         11.4    1.0      0.5        4.10   

        tolls_amount  improvement_surcharge  total_amount  \
190481           0.0                    0.3         12.30   
183318           0.0                    0.3         30.10   
220733           0.0                    0.3         16.30   
5181             0.0                    0.3         20.80   
9816             0.0                    0.3         15.36   
...              ...                    ...           ...   
226656           0.0                    1.0         11.50   
235960           0.0                    1.0         24.72   
228111           0.0                    1.0         12.96   
232115           0.0                    1.0         35.30   
225948           0.0                    1.0         20.50   

        congestion_surcharge  airport_fee  
190481                   2.5          0.0  
183318                   0.0          0.0  
220733                   2.5          0.0  
5181                     2.5          0.0  
9816                     2.5          0.0  
...                      ...          ...  
226656                   2.5          0.0  
235960                   2.5          0.0  
228111                   2.5          0.0  
232115                   2.5          0.0  
225948                   2.5          0.

In [25]:
# Check the count of rows if 10,000 rows were exactly selected or not.
df['tpep_pickup_datetime'].groupby([df['tpep_pickup_datetime'].dt.month]).agg('count')

tpep_pickup_datetime
1     20000
2     20000
3     20000
4     20000
5     20000
6     20000
7     20000
8     20000
9     20000
10    20000
11    20000
12    20000
Name: tpep_pickup_datetime, dtype: int64

In [26]:
df.to_csv('data_2022.csv', index=False)

### 2021 Dataset CSV

In [27]:
df1 = pd.read_parquet('2021.parquet.gzip').dropna(subset=['passenger_count', 'RatecodeID'], how='all')
df1 = df1.groupby([df1['tpep_pickup_datetime'].dt.month]).apply(lambda x: x.sample(20000)).reset_index(drop=True).sort_values(by='tpep_pickup_datetime')
df1

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
16330          2  2009-01-01 12:34:19   2009-01-02 08:22:39              1.0   
2979           1  2021-01-01 00:00:36   2021-01-01 00:19:57              1.0   
1147           2  2021-01-01 00:08:19   2021-01-01 00:28:02              1.0   
16178          2  2021-01-01 00:12:36   2021-01-01 00:22:33              2.0   
17210          1  2021-01-01 00:14:54   2021-01-01 00:18:40              1.0   
...          ...                  ...                   ...              ...   
238040         2  2021-12-31 23:40:48   2021-12-31 23:41:23              1.0   
232051         1  2021-12-31 23:41:30   2022-01-01 00:08:32              3.0   
229026         1  2021-12-31 23:47:36   2022-01-01 00:00:16              2.0   
224770         2  2021-12-31 23:49:46   2021-12-31 23:57:03              1.0   
238867         1  2021-12-31 23:57:12   2022-01-01 00:14:28              1.0   

        trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
16330            1.96         1.0                  N           229   
2979             9.40         1.0                  N           138   
1147             5.47         1.0                  N           238   
16178            2.31         1.0                  N           237   
17210            1.20         1.0                  N           141   
...               ...         ...                ...           ...   
238040           0.06         1.0                  N           186   
232051          11.00         1.0                  N           132   
229026           2.50         1.0                  N           246   
224770           2.17         1.0                  N           141   
238867           9.50         1.0                  N           132   

        DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
16330            264             1         10.0   0.00      0.5        2.66   
2979              90             2         27.5   3.00      0.5        0.00   
1147              79             2         18.5   0.50      0.5        0.00   
16178            164             1         10.0   0.50      0.5        2.76   
17210            236             1          5.5   3.00      0.5        2.00   
...              ...           ...          ...    ...      ...         ...   
238040           186             3         -2.5  -0.50     -0.5        0.00   
232051            61             1         32.5   1.75      0.5        7.00   
229026           125             1         10.5   3.00      0.5        2.00   
224770           107             1          8.5   0.50      0.5        2.00   
238867            39             2         27.5   1.75      0.5        0.00   

        tolls_amount  improvement_surcharge  total_amount  \
16330           0.00                    0.3         15.96   
2979            6.12                    0.3         37.42   
1147            0.00                    0.3         22.30   
16178           0.00                    0.3         16.56   
17210           0.00                    0.3         11.30   
...              ...                    ...           ...   
238040          0.00                   -0.3         -6.30   
232051          0.00                    0.3         42.05   
229026          0.00                    0.3         16.30   
224770          0.00                    0.3         14.30   
238867          0.00                    0.3         30.05   

        congestion_surcharge  airport_fee  
16330                    2.5         0.00  
2979                     2.5          NaN  
1147                     2.5          NaN  
16178                    2.5          NaN  
17210                    2.5          NaN  
...                      ...          ...  
238040                  -2.5         0.00  
232051                   0.0         1.25  
229026                   2.5         0.00  
224770                   2.5         0.00  
238867                   0.0         1.2

In [28]:
# Check the count of rows if 10,000 rows were exactly  selected or not.
df1['tpep_pickup_datetime'].groupby([df1['tpep_pickup_datetime'].dt.month]).agg('count')

tpep_pickup_datetime
1     20000
2     20000
3     20000
4     20000
5     20000
6     20000
7     20000
8     20000
9     20000
10    20000
11    20000
12    20000
Name: tpep_pickup_datetime, dtype: int64

In [29]:
df1.to_csv('data_2021.csv', index=False)

### 2020 Dataset CSV

In [30]:
df2 = pd.read_parquet('2020.parquet.gzip').dropna(subset=['passenger_count', 'RatecodeID'], how='all')
df2 = df2.groupby([df2['tpep_pickup_datetime'].dt.month]).apply(lambda x: x.sample(20000)).reset_index(drop=True).sort_values(by='tpep_pickup_datetime')
df2

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
229856         2  2008-12-31 23:08:07   2008-12-31 23:08:12              1.0   
226182         2  2019-12-31 15:33:38   2019-12-31 15:38:18              1.0   
233881         2  2019-12-31 16:01:57   2019-12-31 16:10:58              1.0   
231399         2  2019-12-31 23:58:55   2020-01-01 00:04:07              2.0   
3015           2  2020-01-01 00:02:21   2020-01-01 00:22:50              1.0   
...          ...                  ...                   ...              ...   
235889         2  2020-12-31 23:43:47   2020-12-31 23:54:04              2.0   
220342         1  2020-12-31 23:45:29   2020-12-31 23:55:27              1.0   
232149         2  2020-12-31 23:47:43   2020-12-31 23:51:21              1.0   
224331         2  2020-12-31 23:48:27   2021-01-01 00:02:08              1.0   
238759         2  2020-12-31 23:59:33   2021-01-01 00:01:57              1.0   

        trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
229856           0.00         1.0                  N           193   
226182           0.62         5.0                  N           114   
233881           1.38         1.0                  N           170   
231399           0.30         1.0                  N           162   
3015             4.69         1.0                  N           125   
...               ...         ...                ...           ...   
235889           2.77         1.0                  N           142   
220342           1.50         1.0                  N            48   
232149           1.07         1.0                  N            90   
224331           6.12         1.0                  N           261   
238759           0.37         1.0                  N            48   

        DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
229856           193             2          2.5    0.0      0.5        0.00   
226182           125             2         60.0    0.0      0.0        0.00   
233881            90             1          8.0    0.5      0.5        2.95   
231399           161             1          5.0    0.5      0.5        2.20   
3015             181             2         17.5    0.5      0.5        0.00   
...              ...           ...          ...    ...      ...         ...   
235889            90             2         10.5    0.5      0.5        0.00   
220342           164             2          8.5    3.0      0.5        0.00   
232149           246             2          5.0    0.5      0.5        0.00   
224331           161             1         19.5    0.5      0.5        5.82   
238759            48             2          3.5    0.5      0.5        0.00   

        tolls_amount  improvement_surcharge  total_amount  \
229856           0.0                    0.3          3.30   
226182           0.0                    0.3         62.80   
233881           0.0                    0.3         14.75   
231399           0.0                    0.3         11.00   
3015             0.0                    0.3         21.30   
...              ...                    ...           ...   
235889           0.0                    0.3         14.30   
220342           0.0                    0.3         12.30   
232149           0.0                    0.3          8.80   
224331           0.0                    0.3         29.12   
238759           0.0                    0.3          7.30   

        congestion_surcharge  airport_fee  
229856                   0.0          NaN  
226182                   2.5          NaN  
233881                   2.5          NaN  
231399                   2.5          NaN  
3015                     2.5          NaN  
...                      ...          ...  
235889                   2.5          NaN  
220342                   2.5          NaN  
232149                   2.5          NaN  
224331                   2.5          NaN  
238759                   2.5          Na

In [31]:
# Check the count of rows if 10,000 rows were exactly  selected or not.
df2['tpep_pickup_datetime'].groupby([df2['tpep_pickup_datetime'].dt.month]).agg('count')

tpep_pickup_datetime
1     20000
2     20000
3     20000
4     20000
5     20000
6     20000
7     20000
8     20000
9     20000
10    20000
11    20000
12    20000
Name: tpep_pickup_datetime, dtype: int64

In [32]:
df2.to_csv('data_2020.csv', index=False)

<hr>

**SUMMARY FILE SIZE:** <br>
1. `data_2022.csv` : 24 MB <br>
2. `data_2021.csv` : 24 MB <br>
3. `data_2020.csv` : 23 MB

## Miscs

In [ ]:
# 2019 and 2018 are not used for this project
# df_2019 = create_df_year(2019)
# df_2018 = create_df_year(2018)

In [ ]:
# Reading data using pandas read_parquet()
# pd.read_parquet('2021.parquet.gzip')